ChemistryLib/PhreeqcIO.cpp

```diff
@@ -301,6 +323,41 @@ std::ostream& operator<<(std::ostream& os, PhreeqcIO const& phreeqc_io)
             os << "-steps " << phreeqc_io._dt << "\n" << "\n";
         }
 
+        os << "USE solution none" << "\n";
+        os << "END" << "\n" << "\n";
+
+        auto const& surface = phreeqc_io._surface;
+        if (!surface.empty())
+        {
+                os << "SURFACE " << chemical_system_id + 1 << "\n";
+                std::size_t aqueous_solution_id =
+                    dump->aqueous_solutions_prev.empty()
+                        ? chemical_system_id + 1
+                        : num_chemical_systems + chemical_system_id + 1;
+                os << "-equilibrate with solution " << aqueous_solution_id << "\n";
+                os << "-sites_units DENSITY" << "\n";
+                os << surface << "\n";
+                os << "END" << "\n" << "\n";
+        }
+
+        os << "USE solution " << chemical_system_id + 1 << "\n";
+
+        if (!equilibrium_phases.empty())
+        {
+            os << "USE equilibrium_phase " << chemical_system_id + 1 << "\n";
+        }
+
+        if (!kinetic_reactants.empty())
+        {
+            os << "USE kinetics " << chemical_system_id + 1 << "\n";
+        }
+
+        if (!surface.empty())
+        {
+            os << "USE surface " << chemical_system_id + 1 << "\n";
+            os << "SAVE solution " << chemical_system_id + 1 << "\n";
+        }
+
```

Comment:
Here you define the Phreeqc inputfile keyword blocs and call them later.
You could shorten this by:

* define `aqueous_solution`
* `if (dump)` define `aqueous_solution_prev`
* `USE solution none` 
* `END`
* `USE solution chemical_system_id + 1`
* `if (!equilibrium_phases.empty())` define `equilibrium_phase`
* `if (!kinetic_reactants.empty())`define `kinetics`
* `if (!surface.empty())` define `surface chemical_system_id + 1` (equilibrate with  aqueous_solution_prev)
* `END`

Reply:
Very helpful. Improved as suggested.